File: lib/nearby/pet-policy.ts
/**
 * ペット同伴可否の共通述語・共通文言。
 *
 * 「店内OK」「テラスOK」フィルタの母集団は確認済みフラグのみ。
 * unknown（null/undefined）は絶対に含めない — 現地で入れなかった、という体験を
 * うちの子と一緒に味わわせないための厳格運用。
 * 地図・検索タブ・スポット詳細の3面でこのモジュールだけを真実として共有する。
 *
 * 判定の真実源（優先順）:
 * 1. pet_indoor_allowed / pet_terrace_only（検証バッチが status と整合させて書く）
 * 2. pet_friendly_status の決定的な値（outdoor_only / not_allowed）
 *
 * 店内OK = 店内（屋内席）への同伴が確認済み
 * テラスOK = テラス席・屋外席での同伴が確認済み（店内不可・外席のみを含む）
 */

/** ペット可否フィールド（共通コントラクト名）を持ちうるスポット */
export type PetPolicySource = {
  pet_indoor_allowed?: boolean | null
  pet_terrace_only?: boolean | null
  pet_friendly_status?: string | null
  pet_friendly_verified?: boolean | null
}

/** 「店内OK」フィルタの表示語（地図・検索タブで共通）。
 *  ラベルは短く保ち、「確認済みのみ」であることはスポット詳細バッジと空状態文言で担保する */
export const INDOOR_OK_FILTER_LABEL = '店内OK'

/** 「テラスOK」フィルタの表示語 */
export const TERRACE_OK_FILTER_LABEL = 'テラスOK'

/** 確認済みで店内OKのスポットか（unknown は必ず false） */
export function placeIsIndoorPetOk(p: PetPolicySource): boolean {
  // 明示 true のみ。status=allowed でも indoor 未確認（null）は通さない
  if (p.pet_indoor_allowed === true) return true
  // 矛盾データ防衛: outdoor_only / not_allowed は店内OKにしない
  return false
}

/**
 * テラス席・屋外席で同伴できる確認があるスポットか。
 * - pet_terrace_only === true（テラス/外席限定の確認済み）
 * - pet_friendly_status === 'outdoor_only'（ステータス側の屋外同伴確認。フラグ欠落のフォールバック）
 * unknown・店内のみ可は含めない。
 */
export function placeIsTerracePetOk(p: PetPolicySource): boolean {
  if (p.pet_friendly_status === 'not_allowed') return false
  if (p.pet_terrace_only === true) return true
  // 検証済み outdoor_only はテラス/屋外同伴の確定シグナル（フラグ未整備のレガシー行を救済）
  if (p.pet_friendly_status === 'outdoor_only') return true
  return false
}

/** スポット詳細の同伴可否バッジ。tone は表示色の使い分け用 */
export type PetPolicyBadge = {
  label: string
  tone: 'ok' | 'terrace' | 'caution'
}

/**
 * 同伴可否バッジの内容を決める。
 *
 * 分岐が indoor / terrace / not_allowed の3つしかなく、leashed_only と
 * 「allowed だが屋内可否は未確認」が揃って null に落ちていた。実データで測ると
 * 70%（50件中35件）がバッジ無しで、「リード着用なら同伴可」と
 * 「そもそも誰も調べていない」が画面上で見分けられなかった。
 *
 * 検証済みで分かっていることは、弱い情報でも必ず何か出す。出さないのは
 * 本当に未検証のときだけにする。
 */
export function petPolicyBadge(p: PetPolicySource): PetPolicyBadge | null {
  if (p.pet_friendly_status === 'not_allowed') return { label: '同伴不可の可能性', tone: 'caution' }
  /*
    「確認済み」は名乗らない。

    pet_friendly_verified が立っていても、その根拠が推測のものが実測で 2,602件（確定
    ステータスの16.1%）ある。注記に「推測」「要確認」「web検索は実施できず」と自分で
    書き残しながら確定値になっている行が、そのまま同じバッジで出ていた。

    根拠の質を伴わない「確認済み」は、無根拠の断定より悪い。飼い主はそう書かれた
    情報を最も強く信じる。根拠の質を返す仕組みができるまでは、可否だけを述べる。
  */
  if (p.pet_indoor_allowed === true) return { label: '店内OK', tone: 'ok' }
  if (placeIsTerracePetOk(p)) return { label: 'テラス席のみOK', tone: 'terrace' }
  if (p.pet_friendly_status === 'leashed_only') return { label: 'リード着用で同伴OK', tone: 'ok' }
  // 同伴自体は確認済みだが、屋内まで入れるかは分かっていない。
  // 「OK」だけを出すと店内に入れる前提で来てしまうので、そこは言い切らない
  if (p.pet_friendly_status === 'allowed') return { label: '同伴OK・店内は要確認', tone: 'ok' }
  // 検証は済んでいるが可否を判定できなかった。未検証（何も出さない）とは別の状態
  if (p.pet_friendly_verified === true) return { label: '同伴可否は要確認', tone: 'caution' }
  return null
}

File: swift/WanspotKit/Sources/WanspotKit/Domain/Nearby/PetPolicy.swift
import Foundation

public protocol PetPolicySource: Sendable {
    var petIndoorAllowed: Bool? { get }
    var petTerraceOnly: Bool? { get }
    var petFriendlyStatus: String? { get }
    var petFriendlyVerified: Bool? { get }
    var petPolicyEvidence: String? { get }
    var petSizeLimit: String? { get }
    var petReservationRequired: Bool? { get }
    var dogInteraction: String? { get }
}

extension PlaceResult: PetPolicySource {}
extension SpotDetail: PetPolicySource {}

public enum PetPolicyEvidenceTone: String, Codable, Equatable, Sendable {
    case confirmed
    case reported
    case weak
}

public enum PetPolicyBadgeTone: String, Codable, Equatable, Sendable {
    case ok
    case terrace
    case caution
}

public struct PetPolicyBadge: Codable, Equatable, Sendable {
    public let label: String
    public let tone: PetPolicyBadgeTone

    public init(label: String, tone: PetPolicyBadgeTone) {
        self.label = label
        self.tone = tone
    }
}

public enum PetAccessConditionKind: String, Codable, Equatable, Sendable {
    case size
    case reservation
    case interaction
}

public struct PetAccessCondition: Codable, Equatable, Sendable {
    public let kind: PetAccessConditionKind
    public let label: String
    public let isCaution: Bool

    public init(
        kind: PetAccessConditionKind,
        label: String,
        isCaution: Bool = false
    ) {
        self.kind = kind
        self.label = label
        self.isCaution = isCaution
    }
}

public struct PetPolicyPresentation: Equatable, Sendable {
    public let badge: PetPolicyBadge?
    public let conditions: [PetAccessCondition]
    public let advisory: String?

    public init(
        badge: PetPolicyBadge?,
        conditions: [PetAccessCondition],
        advisory: String?
    ) {
        self.badge = badge
        self.conditions = conditions
        self.advisory = advisory
    }
}

public enum PetPolicy {
    public static let indoorFilterLabel = "店内OK"
    public static let terraceFilterLabel = "テラスOK"

    public static func evidenceTone<Source: PetPolicySource>(
        for spot: Source
    ) -> PetPolicyEvidenceTone {
        switch spot.petPolicyEvidence {
        case "official", "aggregator":
            .confirmed
        case "reviews":
            .reported
        default:
            .weak
        }
    }

    public static func isIndoorAllowed<Source: PetPolicySource>(
        _ spot: Source
    ) -> Bool {
        spot.petIndoorAllowed == true
    }

    public static func isTerraceAllowed<Source: PetPolicySource>(
        _ spot: Source
    ) -> Bool {
        guard spot.petFriendlyStatus != "not_allowed" else { return false }
        return spot.petTerraceOnly == true
            || spot.petFriendlyStatus == "outdoor_only"
    }

    public static func badge<Source: PetPolicySource>(
        for spot: Source
    ) -> PetPolicyBadge? {
        if spot.petFriendlyStatus == "not_allowed" {
            return PetPolicyBadge(
                label: "同伴不可の可能性",
                tone: .caution
            )
        }
        if spot.petIndoorAllowed == true {
            return PetPolicyBadge(
                label: labelWithEvidence("店内OK", spot: spot),
                tone: .ok
            )
        }
        if isTerraceAllowed(spot) {
            return PetPolicyBadge(
                label: labelWithEvidence("テラス席のみOK", spot: spot),
                tone: .terrace
            )
        }
        if spot.petFriendlyStatus == "leashed_only" {
            return PetPolicyBadge(
                label: labelWithEvidence("リード着用で同伴OK", spot: spot),
                tone: .ok
            )
        }
        if spot.petFriendlyStatus == "allowed" {
            return PetPolicyBadge(
                label: labelWithEvidence(
                    "同伴OK・店内は要確認",
                    spot: spot
                ),
                tone: .ok
            )
        }
        if spot.petFriendlyVerified == true {
            return PetPolicyBadge(
                label: "同伴可否は要確認",
                tone: .caution
            )
        }
        return nil
    }

    public static func presentation<Source: PetPolicySource>(
        for spot: Source
    ) -> PetPolicyPresentation {
        var conditions: [PetAccessCondition] = []
        if
            let limit = spot.petSizeLimit?
                .trimmingCharacters(in: .whitespacesAndNewlines),
            !limit.isEmpty
        {
            conditions.append(
                PetAccessCondition(
                    kind: .size,
                    label: "サイズ条件: \(limit)",
                    isCaution: true
                )
            )
        }
        if spot.petReservationRequired == true {
            conditions.append(
                PetAccessCondition(
                    kind: .reservation,
                    label: "愛犬同伴での利用は予約が必要です",
                    isCaution: true
                )
            )
        }
        if spot.dogInteraction == "meet_dogs" {
            conditions.append(
                PetAccessCondition(
                    kind: .interaction,
                    label: "お店のワンちゃんと触れ合う施設です。愛犬の同伴可否は事前にご確認ください",
                    isCaution: true
                )
            )
        }

        let badge = badge(for: spot)
        let advisory: String?
        if badge == nil {
            advisory = "同伴条件はまだ確認できていません。お出かけ前に施設へご確認ください。"
        } else if spot.petFriendlyStatus == "not_allowed" {
            advisory = "施設の最新ルールが変わっている場合があります。来訪前にご確認ください。"
        } else {
            advisory = nil
        }
        return PetPolicyPresentation(
            badge: badge,
            conditions: conditions,
            advisory: advisory
        )
    }

    private static func labelWithEvidence<Source: PetPolicySource>(
        _ base: String,
        spot: Source
    ) -> String {
        switch evidenceTone(for: spot) {
        case .confirmed:
            "\(base)・確認済み"
        case .reported:
            "\(base)（口コミによる）"
        case .weak:
            base
        }
    }
}
